applications/StructuralMechanicsApplication/python_scripts/structural_response.py

```diff
@@ -220,7 +220,215 @@ def GetShapeGradient(self):
+        self.adjoint_analysis = structural_mechanics_analysis.StructuralMechanicsAnalysis(adjoint_model, ProjectParametersAdjoint)
+        self.adjoint_model_part_name = ProjectParametersAdjoint["problem_data"]["model_part_name"].GetString()
+
+
+    def Initialize(self):
+        self.primal_analysis.Initialize()
+        self.adjoint_analysis.Initialize()
+
+        self.primal_model_part = self.primal_analysis.model.GetModelPart(self.primal_model_part_name)
+        self.adjoint_model_part = self.adjoint_analysis.model.GetModelPart(self.adjoint_model_part_name)
+
+        # TODO should be created here, not in solver!
+        self.response_function_utility = self.adjoint_analysis._GetSolver().response_function
+
+
+    def InitializeSolutionStep(self):
+        # synchronize the modelparts # TODO this should happen automatically
+        print("\n> Synchronize primal and adjoint modelpart for response:", self.identifier)
+        for node in self.adjoint_model_part.Nodes:
+            ref_node = self.primal_model_part.Nodes[node.Id]
+            node.X0 = ref_node.X0
+            node.Y0 = ref_node.Y0
+            node.Z0 = ref_node.Z0
+            node.X = ref_node.X
+            node.Y = ref_node.Y
+            node.Z = ref_node.Z
+
+
+        # Run the primal analysis.
+        # TODO if primal_analysis.status==solved: return
+        print("\n> Starting primal analysis for response:", self.identifier)
+        startTime = timer.time()
+        self.primal_analysis.time = self.primal_analysis._GetSolver().AdvanceInTime(self.primal_analysis.time)
+        self.primal_analysis.InitializeSolutionStep()
+        self.primal_analysis._GetSolver().Predict()
+        self.primal_analysis._GetSolver().SolveSolutionStep()
+        self.primal_analysis.FinalizeSolutionStep()
+        self.primal_analysis.OutputSolutionStep()
+        print("> Time needed for solving the primal analysis = ",round(timer.time() - startTime,2),"s")
+
+        # TODO the response value calculation for stresses currently only works on the adjoint modelpart
+        # this needs to be improved, also the response value should be calculated on the PRIMAL modelpart!!
+        self.adjoint_analysis.time = self.adjoint_analysis._GetSolver().AdvanceInTime(self.adjoint_analysis.time)
+        self.adjoint_analysis.InitializeSolutionStep()
+
+    def CalculateValue(self):
+        startTime = timer.time()
+        value = self.response_function_utility.CalculateValue(self.adjoint_model_part)
+        print("> Time needed for calculating the response value = ",round(timer.time() - startTime,2),"s")
+
+        self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.RESPONSE_VALUE] = value
+
+
+    def CalculateGradient(self):
+        print("\n> Starting adjoint analysis for response:", self.identifier)
+        startTime = timer.time()
+        self.adjoint_analysis._GetSolver().Predict()
+        self.adjoint_analysis._GetSolver().SolveSolutionStep()
```

Comment:
Can we replace these with RunSolutionLoop()?

Reply:
See comment above

Reply:
Still not resolved. At least call `InitializeSolutionStep`.

Reply:
This is part of the requested changes of the current review.

Reply:
`adjoint_analysis.InitializeSolutionStep` and `adjoint_analysis.FinalizeSolutionStep` are now called in the respective functions of the AdjointResponseFunction (python class)